File: go/sort/merge.go
package sort

// MergeSort 对整数切片进行归并排序，返回排序后的新切片
func MergeSort(arr []int) []int {
	n := len(arr)
	if n <= 1 {
		return arr
	}

	mid := n / 2
	left := MergeSort(arr[:mid])
	right := MergeSort(arr[mid:])

	return merge(left, right)
}

// merge 合并两个有序切片
func merge(left, right []int) []int {
	result := make([]int, 0, len(left)+len(right))
	i, j := 0, 0

	for i < len(left) && j < len(right) {
		if left[i] <= right[j] {
			result = append(result, left[i])
			i++
		} else {
			result = append(result, right[j])
			j++
		}
	}

	result = append(result, left[i:]...)
	result = append(result, right[j:]...)
	return result
}

File: cpp/sort/merge.cpp
#include <algorithm>
#include <iostream>
#include <vector>

using namespace std;

void merge_sort(vector<int>& arr, int left, int right) {
    if (left >= right) return;
    const int mid = left + ((right - left) >> 1);
    merge_sort(arr, left, mid);
    merge_sort(arr, mid + 1, right);

    vector<int> tmp;
    tmp.reserve(right - left + 1);
    int i = left, j = mid + 1;
    while (i <= mid && j <= right) {
        if (arr[i] <= arr[j]) tmp.push_back(arr[i++]);
        else                  tmp.push_back(arr[j++]);
    }
    while (i <= mid)   tmp.push_back(arr[i++]);
    while (j <= right) tmp.push_back(arr[j++]);

    std::copy(tmp.begin(), tmp.end(), arr.begin() + left);
}

int main() {
    vector<int> arr = {3, 5, 1, 4, 2};
    if (!arr.empty()) {
        merge_sort(arr, 0, static_cast<int>(arr.size()) - 1);
    }
    for (const int i : arr) cout << i << ' ';
    cout << '\n';
    return 0;
}